### PR TITLE
Port Union algorithm to C++ and fix it

### PR DESCRIPTION
**Pull request by wonder-sk**
Following up on #6853 

This turns the union algorithm into a combination of existing algorithms, lowering the room for incorrectly handled corner cases. The union results are given as `(A intersection B) U (A - B) U (B - A)`

Should fix problems mentioned in #17131 with the first dataset. The second dataset in that ticket is not handled yet.

I had to remove older processing test case for union/difference - besides covering fewer cases than the new overlay1 dataset, the expected union result was even wrong.

With this PR all the basic geometry processing algs are in c++ and covered with basic unit tests.

### CODE REVIEW

File: src/analysis/processing/qgsoverlayutils.h [hunk 1]
```diff
@@ -38,6 +41,11 @@ namespace QgsOverlayUtils
 
   void difference( const QgsFeatureSource &sourceA, const QgsFeatureSource &sourceB, QgsFeatureSink &sink, QgsProcessingContext &context, QgsProcessingFeedback *feedback, int &count, int totalCount, DifferenceOutput outputAttrs );
 
+  void intersection( const QgsFeatureSource &sourceA, const QgsFeatureSource &sourceB, QgsFeatureSink &sink, QgsProcessingContext &context, QgsProcessingFeedback *feedback, int &count, int totalCount, const QList<int> &fieldIndicesA, const QList<int> &fieldIndicesB );
+
+  QList<int> fieldNamesToIndices( const QStringList &fieldNames, const QgsFields &fields );
```
Comment (nyalldawson) [inline src/analysis/processing/qgsoverlayutils.h]: Any change you can move this and indicesToFields to QgsProcessingUtils (and add tests)? There's other (non-overlay) algs which do the same things, and I'd like to reuse this code.

**Review by nyalldawson — APPROVED** [review]

**Comment by nyalldawson** [comment]
Looks great to me! Very welcome improvements!

File: src/analysis/processing/qgsoverlayutils.h (hunk 1 above)
Comment (wonder-sk) [reply src/analysis/processing/qgsoverlayutils.h]: Sure - moved in 091ed91 and tests added in 5e7719c

**Comment by nyalldawson** [comment]
Thanks!

**Pull request merged**